### PR TITLE
Deprecate existing coupon codes methods

### PR DESCRIPTION
**Pull request by aitbw**
Following @kennyadsl's idea on solidusio#2327, this PR aims to deprecate current endpoints used to apply coupon codes (on orders and checkout) and replace them with a new API controller/endpoint properly named `Spree::Api::CouponCodesController`.

I think this approach allows a better separation of concerns and it's easier to reason about. Plus, it'd be easier to implement a built-in solution for solidusio#1641.

That said, we're still missing a `Spree::CouponCodesController` controller/endpoint under `/frontend` but since we're already using API calls under said folder, I still debating whether to implement that or not.

I'm open to suggestions!

### CODE REVIEW

File: api/spec/features/checkout_spec.rb [hunk 1]
```diff
@@ -61,6 +61,8 @@ def create_line_item(variant, quantity = 1)
     end
 
     def add_promotion(_promotion)
+      expect(Spree::Deprecation).to receive(:warn)
```
Comment (kennyadsl) [inline api/spec/features/checkout_spec.rb]: what about use the new api endpoint in the spec instead of keeping the old one and expecting a deprecation warning?

File: api/spec/requests/spree/api/promotion_application_spec.rb [hunk 2]
```diff
@@ -20,6 +20,7 @@ module Spree::Api
 
       it "can apply a coupon code to the order" do
```
Comment (kennyadsl) [inline api/spec/requests/spree/api/promotion_application_spec.rb]: this tests look like duplicated of what we have [here](https://github.com/nebulab/solidus/pull/19/files#diff-72468e52a59aedfe8a3443597f1a6f3aR856), right?  If yes, maybe we can do another PR to remove duplication?

File: api/spec/features/checkout_spec.rb (hunk 1 above)
Comment (aitbw) [reply api/spec/features/checkout_spec.rb]: Sure!

File: api/spec/requests/spree/api/promotion_application_spec.rb (hunk 2 above)
Comment (aitbw) [reply api/spec/requests/spree/api/promotion_application_spec.rb]: Noted :ok_hand:

File: api/spec/requests/spree/api/promotion_application_spec.rb [hunk 3]
```diff
@@ -20,6 +20,7 @@ module Spree::Api
...
         expect(order.total).to eq(110.00)
+        expect(Spree::Deprecation).to receive(:warn)
```
Comment (kennyadsl) [inline api/spec/requests/spree/api/promotion_application_spec.rb]: I'm not sure to understand why specs in these files are needed, maybe we can move them into `api/spec/requests/spree/api/coupon_codes_controller_spec.rb` ? Of course we should also remove the `expect(Spree::Deprecation).to receive(:warn)` and make them use the new endpoint. WDYT?

**Comment by kennyadsl** [comment]
Also, we can maybe squash some commits together?

File: api/spec/requests/spree/api/promotion_application_spec.rb (hunk 3 above)
Comment (aitbw) [reply api/spec/requests/spree/api/promotion_application_spec.rb]: Sounds good to me! Although I'm assuming we'd need another PR to remove duplication, am I right? Like in [here](https://github.com/nebulab/solidus/pull/19#discussion_r228501278)

Comment (kennyadsl) [reply api/spec/requests/spree/api/promotion_application_spec.rb]: Makes perfect sense!

**Comment by aitbw** [comment]
Closing now that solidusio#2958 is merged into master
